File: go-nanopdf/examples/easy_api_examples.go
// Package main provides examples of using the NanoPDF Easy API
//
// The Easy API provides a simplified, ergonomic interface for common PDF tasks
// with automatic resource management and intuitive method chaining.
package main

import (
	"fmt"
	"log"
	"os"

	"github.com/lexmata/nanopdf/go-nanopdf"
	"github.com/lexmata/nanopdf/go-nanopdf/easy"
)

// Example1_SimpleTextExtraction demonstrates the simplest way to extract text
func Example1_SimpleTextExtraction() {
	// Extract all text
	text, err := easy.ExtractText("document.pdf")
	if err != nil {
		log.Fatal(err)
	}
	fmt.Println(text)

	// Extract from specific page
	pageText, err := easy.ExtractPageText("document.pdf", 0)
	if err != nil {
		log.Fatal(err)
	}
	fmt.Println(pageText)

	// Save text to file
	err = easy.SaveTextToFile("document.pdf", "output.txt")
	if err != nil {
		log.Fatal(err)
	}
}

// Example2_SimpleRendering demonstrates the simplest way to render pages
func Example2_SimpleRendering() {
	// Render first page at 150 DPI
	err := easy.RenderToPNG("document.pdf", "output.png", 0, 150)
	if err != nil {
		log.Fatal(err)
	}

	// Render specific page at high DPI
	err = easy.RenderToPNG("document.pdf", "output.png", 0, 300)
	if err != nil {
		log.Fatal(err)
	}

	// Render all pages
	err = easy.RenderAllToPNG("document.pdf", "output/page-{page}.png", 150)
	if err != nil {
		log.Fatal(err)
	}
}

// Example3_SimpleInfo demonstrates getting PDF information
func Example3_SimpleInfo() {
	// Get page count
	pageCount, err := easy.GetPageCount("document.pdf")
	if err != nil {
		log.Fatal(err)
	}
	fmt.Printf("Pages: %d\n", pageCount)

	// Get metadata
	metadata, err := easy.GetMetadata("document.pdf")
	if err != nil {
		log.Fatal(err)
	}
	fmt.Printf("Title: %s\n", metadata.Title)
	fmt.Printf("Author: %s\n", metadata.Author)

	// Get complete info
	info, err := easy.GetInfo("document.pdf")
	if err != nil {
		log.Fatal(err)
	}
	fmt.Printf("Document has %d pages\n", info.PageCount)

	// Quick summary
	summary, err := easy.QuickSummary("document.pdf")
	if err != nil {
		log.Fatal(err)
	}
	fmt.Println(summary)
}

// Example4_SimpleSearch demonstrates text searching
func Example4_SimpleSearch() {
	// Find pages containing text
	pages, err := easy.Search("document.pdf", "important")
	if err != nil {
		log.Fatal(err)
	}
	fmt.Printf("Found on pages: %v\n", pages)

	// Check if encrypted
	encrypted, err := easy.IsEncrypted("document.pdf")
	if err != nil {
		log.Fatal(err)
	}
	if encrypted {
		fmt.Println("PDF is password protected")
	}
}

// Example5_FluentAPI demonstrates the fluent builder API
func Example5_FluentAPI() {
	// Open and use with automatic cleanup
	pdf, err := easy.Open("document.pdf")
	if err != nil {
		log.Fatal(err)
	}
	defer pdf.Close()

	pageCount := pdf.PageCount()
	metadata := pdf.GetMetadata()
	text, _ := pdf.ExtractAllText()

	fmt.Printf("Pages: %d\n", pageCount)
	fmt.Printf("Title: %s\n", metadata.Title)
	fmt.Printf("Text length: %d characters\n", len(text))
}

// Example6_ManualResourceManagement demonstrates manual cleanup
func Example6_ManualResourceManagement() {
	pdf, err := easy.Open("document.pdf").KeepOpen()
	if err != nil {
		log.Fatal(err)
	}
	defer pdf.Close()

	// Extract text from multiple pages
	for i := 0; i < pdf.PageCount(); i++ {
		text, _ := pdf.ExtractPageText(i)
		fmt.Printf("Page %d: %s...\n", i, text[:min(100, len(text))])
	}

	// Search across document
	results, _ := pdf.Search("important", -1)
	fmt.Printf("Found %d occurrences\n", len(results))
}

// Example7_AdvancedRendering demonstrates advanced rendering options
func Example7_AdvancedRendering() {
	pdf, err := easy.Open("document.pdf")
	if err != nil {
		log.Fatal(err)
	}
	defer pdf.Close()

	// Render at specific DPI
	err = pdf.RenderToFile(0, "high-quality.png", easy.RenderOptions{
		DPI:    300,
		Format: "png",
	})
	if err != nil {
		log.Fatal(err)
	}

	// Render with specific dimensions
	err = pdf.RenderToFile(0, "thumbnail.png", easy.RenderOptions{
		Width:  200,
		Height: 300,
		Format: "png",
	})
	if err != nil {
		log.Fatal(err)
	}

	// Render in grayscale
	err = pdf.RenderToFile(0, "grayscale.png", easy.RenderOptions{
		DPI:        150,
		Colorspace: "gray",
		Format:     "png",
	})
	if err != nil {
		log.Fatal(err)
	}

	// Render with alpha channel
	err = pdf.RenderToFile(0, "with-alpha.png", easy.RenderOptions{
		DPI:    150,
		Alpha:  true,
		Format: "png",
	})
	if err != nil {
		log.Fatal(err)
	}
}

// Example8_DetailedSearch demonstrates search with detailed results
func Example8_DetailedSearch() {
	pdf, err := easy.Open("document.pdf")
	if err != nil {
		log.Fatal(err)
	}
	defer pdf.Close()

	results, err := pdf.Search("important", -1)
	if err != nil {
		log.Fatal(err)
	}

	fmt.Printf("Found %d occurrences:\n", len(results))
	for i, result := range results {
		fmt.Printf("%d. Page %d:\n", i+1, result.PageNumber)
		fmt.Printf("   Text: \"%s\"\n", result.Text)
		fmt.Printf("   Position: (%.2f, %.2f)\n", result.BBox.X0, result.BBox.Y0)
		fmt.Printf("   Size: %.2fx%.2f\n", result.BBox.Width(), result.BBox.Height())
	}
}

// Example9_CompleteInfo demonstrates getting complete document information
func Example9_CompleteInfo() {
	pdf, err := easy.Open("document.pdf")
	if err != nil {
		log.Fatal(err)
	}
	defer pdf.Close()

	info := pdf.GetInfo()

	fmt.Println("Document Information:")
	fmt.Printf("  Pages: %d\n", info.PageCount)
	fmt.Printf("  Encrypted: %v\n", info.IsEncrypted)

	fmt.Println("\nMetadata:")
	if info.Metadata.Title != "" {
		fmt.Printf("  Title: %s\n", info.Metadata.Title)
	}
	if info.Metadata.Author != "" {
		fmt.Printf("  Author: %s\n", info.Metadata.Author)
	}
	if info.Metadata.Subject != "" {
		fmt.Printf("  Subject: %s\n", info.Metadata.Subject)
	}

	fmt.Println("\nPages:")
	for _, page := range info.Pages {
		fmt.Printf("  Page %d: %.0fx%.0f (rotation: %d°)\n",
			page.PageNumber, page.Width, page.Height, page.Rotation)
	}
}

// Example10_BatchProcessing demonstrates processing multiple PDFs
func Example10_BatchProcessing() {
	files := []string{"doc1.pdf", "doc2.pdf", "doc3.pdf"}

	for _, file := range files {
		fmt.Printf("\nProcessing %s...\n", file)

		// Extract text
		err := easy.SaveTextToFile(file, file+".txt")
		if err != nil {
			log.Printf("Error extracting text: %v", err)
			continue
		}

		// Render first page
		err = easy.RenderToPNG(file, file+".png", 0, 150)
		if err != nil {
			log.Printf("Error rendering: %v", err)
			continue
		}

		// Get info
		pageCount, _ := easy.GetPageCount(file)
		fmt.Printf("  %d pages processed\n", pageCount)
	}
}

// Example11_ErrorHandling demonstrates error handling
func Example11_ErrorHandling() {
	// Try to open encrypted PDF without password
	_, err := easy.ExtractText("encrypted.pdf")
	if err != nil {
		fmt.Printf("Error: %v\n", err)

		// Try again with password
		pdf, err := easy.OpenWithPassword("encrypted.pdf", "password123")
		if err != nil {
			fmt.Printf("Wrong password: %v\n", err)
			return
		}
		defer pdf.Close()

		fmt.Println("Success with password!")
	}
}

// Example12_UseCallback demonstrates the Use callback pattern
func Example12_UseCallback() {
	err := func() error {
		pdf, err := easy.Open("document.pdf")
		if err != nil {
			return err
		}

		return pdf.Use(func(p *easy.PDF) error {
			text, err := p.ExtractAllText()
			if err != nil {
				return err
			}

			fmt.Printf("Extracted %d characters\n", len(text))
			return nil
		})
	}()

	if err != nil {
		log.Fatal(err)
	}
}

// Example13_WorkingWithBytes demonstrates working with byte data
func Example13_WorkingWithBytes() {
	// Read PDF file
	data, err := os.ReadFile("document.pdf")
	if err != nil {
		log.Fatal(err)
	}

	// Open from bytes
	pdf, err := easy.FromBytes(data)
	if err != nil {
		log.Fatal(err)
	}
	defer pdf.Close()

	text, _ := pdf.ExtractAllText()
	pageCount := pdf.PageCount()

	fmt.Printf("Loaded %d pages from buffer\n", pageCount)
	fmt.Printf("Extracted %d characters\n", len(text))
}

// Example14_RenderToMemory demonstrates rendering to memory
func Example14_RenderToMemory() {
	pdf, err := easy.Open("document.pdf")
	if err != nil {
		log.Fatal(err)
	}
	defer pdf.Close()

	// Render first page to bytes
	imageData, err := pdf.RenderToBytes(0, easy.RenderOptions{
		DPI:    300,
		Format: "png",
	})
	if err != nil {
		log.Fatal(err)
	}

	fmt.Printf("Rendered image: %d bytes\n", len(imageData))

	// Save to file
	err = os.WriteFile("output.png", imageData, 0644)
	if err != nil {
		log.Fatal(err)
	}
}

// Example15_RenderAllPages demonstrates rendering all pages
func Example15_RenderAllPages() {
	pdf, err := easy.Open("document.pdf")
	if err != nil {
		log.Fatal(err)
	}
	defer pdf.Close()

	// Render all pages to memory
	images, err := pdf.RenderAll(easy.RenderOptions{
		DPI:    150,
		Format: "png",
	})
	if err != nil {
		log.Fatal(err)
	}

	fmt.Printf("Rendered %d pages\n", len(images))

	// Save each image
	for i, imageData := range images {
		filename := fmt.Sprintf("page-%03d.png", i)
		err := os.WriteFile(filename, imageData, 0644)
		if err != nil {
			log.Printf("Error saving page %d: %v", i, err)
		}
	}
}

// Example16_ComparisonWithCoreAPI demonstrates the difference between Easy and Core APIs
func Example16_ComparisonWithCoreAPI() {
	// Easy API - Simple and clean
	text, _ := easy.ExtractText("document.pdf")
	fmt.Println(text)

	// Core API - More verbose but maximum control
	doc, _ := nanopdf.OpenDocument("document.pdf", "")
	defer doc.Close()

	pageCount := doc.PageCount()
	for i := 0; i < pageCount; i++ {
		page, _ := doc.LoadPage(i)
		pageText := page.ExtractText()
		fmt.Println(pageText)
		page.Drop()
	}
}

func main() {
	fmt.Println("NanoPDF Easy API Examples")
	fmt.Println("==========================")

	// Note: Most examples are functions that you can call individually
	// Uncomment and modify file paths as needed for your use case

	// Example1_SimpleTextExtraction()
	// Example2_SimpleRendering()
	// Example3_SimpleInfo()
	// Example4_SimpleSearch()
	// Example5_FluentAPI()
	// Example6_ManualResourceManagement()
	// Example7_AdvancedRendering()
	// Example8_DetailedSearch()
	// Example9_CompleteInfo()
	// Example10_BatchProcessing()
	// Example11_ErrorHandling()
	// Example12_UseCallback()
	// Example13_WorkingWithBytes()
	// Example14_RenderToMemory()
	// Example15_RenderAllPages()
	// Example16_ComparisonWithCoreAPI()

	fmt.Println("\nAll examples are defined as functions.")
	fmt.Println("Uncomment the ones you want to run in main().")
}

// Helper function
func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

File: nanopdf-js/examples/easy-api-examples.ts
/**
 * NanoPDF Easy API Examples
 *
 * This file demonstrates the ergonomic, user-friendly API for common PDF tasks.
 * Choose the API level that best fits your needs:
 *
 * 1. Simple API - Ultra-simple functions (best for beginners)
 * 2. Easy API - Fluent builder pattern (balance of power and simplicity)
 * 3. Core API - Full control (advanced users)
 */

import * as simple from '../src/simple.js';
import { EasyPDF, PDFUtils } from '../src/easy.js';
import { Document } from '../src/document.js';

// ============================================================================
// SIMPLE API - Ultra-Simple Functions
// ============================================================================

/**
 * Example 1: Extract text from a PDF (simplest way)
 */
async function example1_SimpleTextExtraction() {
  // Extract all text
  const text = await simple.extractText('document.pdf');
  console.log(text);

  // Extract from specific page
  const pageText = await simple.extractPageText('document.pdf', 0);
  console.log(pageText);

  // Save text to file
  await simple.saveTextToFile('document.pdf', 'output.txt');
}

/**
 * Example 2: Render PDF pages to PNG (simplest way)
 */
async function example2_SimpleRendering() {
  // Render first page at 150 DPI
  await simple.renderToPNG('document.pdf', 'output.png');

  // Render specific page at high DPI
  await simple.renderToPNG('document.pdf', 'output.png', 0, 300);

  // Render all pages
  await simple.renderAllToPNG('document.pdf', 'output/page-{page}.png', 150);

  // Get page as buffer
  const buffer = await simple.getPageAsPNG('document.pdf', 0, 300);
  // Do something with buffer...
}

/**
 * Example 3: Get PDF information (simplest way)
 */
async function example3_SimpleInfo() {
  // Get page count
  const pageCount = await simple.getPageCount('document.pdf');
  console.log(`Pages: ${pageCount}`);

  // Get metadata
  const metadata = await simple.getMetadata('document.pdf');
  console.log(`Title: ${metadata.title}`);
  console.log(`Author: ${metadata.author}`);

  // Get complete info
  const info = await simple.getPdfInfo('document.pdf');
  console.log(JSON.stringify(info, null, 2));

  // Quick summary
  const summary = await simple.quickSummary('document.pdf');
  console.log(summary);
}

/**
 * Example 4: Search text (simplest way)
 */
async function example4_SimpleSearch() {
  // Find pages containing text
  const pages = await simple.searchText('document.pdf', 'important');
  console.log(`Found on pages: ${pages.join(', ')}`);

  // Check if encrypted
  if (await simple.isEncrypted('document.pdf')) {
    console.log('PDF is password protected');
  }
}

/**
 * Example 5: Convert PDF to images (simplest way)
 */
async function example5_SimpleConversion() {
  // Convert all pages to PNG images
  const files = await simple.convertToImages('document.pdf', 'output/', {
    dpi: 300,
    format: 'png'
  });

  console.log(`Created ${files.length} images:`);
  files.forEach((file) => console.log(`  ${file}`));
}

// ============================================================================
// EASY API - Fluent Builder Pattern
// ============================================================================

/**
 * Example 6: Fluent API with automatic resource management
 */
async function example6_FluentAPI() {
  // Open and use with automatic cleanup
  await EasyPDF.open('document.pdf').useAsync(async (pdf) => {
    const pageCount = pdf.pageCount;
    const metadata = pdf.getMetadata();
    const text = pdf.extractAllText();

    console.log(`Pages: ${pageCount}`);
    console.log(`Title: ${metadata.title}`);
    console.log(`Text length: ${text.length} characters`);
  });
  // PDF is automatically closed here
}

/**
 * Example 7: Fluent API with manual resource management
 */
function example7_ManualResourceManagement() {
  const pdf = EasyPDF.open('document.pdf').keepOpen();

  try {
    // Extract text from multiple pages
    for (let i = 0; i < pdf.pageCount; i++) {
      const text = pdf.extractPageText(i);
      console.log(`Page ${i}:`, text.substring(0, 100));
    }

    // Search across document
    const results = pdf.search('important');
    console.log(`Found ${results.length} occurrences`);

    // Render pages
    const buffer = pdf.renderToBuffer(0, { dpi: 300 });
    // Do something with buffer...
  } finally {
    pdf.close();
  }
}

/**
 * Example 8: Advanced rendering options
 */
async function example8_AdvancedRendering() {
  await EasyPDF.open('document.pdf').useAsync(async (pdf) => {
    // Render at specific DPI
    await pdf.renderToFile(0, 'high-quality.png', {
      dpi: 300,
      format: 'png'
    });

    // Render with specific dimensions
    await pdf.renderToFile(0, 'thumbnail.png', {
      width: 200,
      height: 300,
      format: 'png'
    });

    // Render in grayscale
    await pdf.renderToFile(0, 'grayscale.png', {
      dpi: 150,
      colorspace: 'gray',
      format: 'png'
    });

    // Render with alpha channel
    await pdf.renderToFile(0, 'with-alpha.png', {
      dpi: 150,
      alpha: true,
      format: 'png'
    });
  });
}

/**
 * Example 9: Structured text extraction with bounding boxes
 */
async function example9_StructuredText() {
  await EasyPDF.open('document.pdf').useAsync(async (pdf) => {
    // Extract structured text from first page
    const structured = pdf.extractStructuredText(0);

    structured.forEach((item) => {
      console.log(`Page ${item.pageNumber}:`);
      console.log(`Text: ${item.text}`);

      if (item.blocks) {
        item.blocks.forEach((block, i) => {
          console.log(`  Block ${i}:`);
          console.log(`    Text: ${block.text}`);
          console.log(`    Position: (${block.bbox.x}, ${block.bbox.y})`);
          console.log(`    Size: ${block.bbox.width}x${block.bbox.height}`);
        });
      }
    });
  });
}

/**
 * Example 10: Search with detailed results
 */
async function example10_DetailedSearch() {
  await EasyPDF.open('document.pdf').useAsync(async (pdf) => {
    const results = pdf.search('important');

    console.log(`Found ${results.length} occurrences:`);
    results.forEach((result, i) => {
      console.log(`${i + 1}. Page ${result.pageNumber}:`);
      console.log(`   Text: "${result.text}"`);
      console.log(`   Position: (${result.bbox.x}, ${result.bbox.y})`);
      console.log(`   Size: ${result.bbox.width}x${result.bbox.height}`);
    });
  });
}

/**
 * Example 11: Extract complete document information
 */
async function example11_CompleteInfo() {
  await EasyPDF.open('document.pdf').useAsync(async (pdf) => {
    const info = pdf.getInfo();

    console.log('Document Information:');
    console.log(`  Pages: ${info.pageCount}`);
    console.log(`  Encrypted: ${info.isEncrypted}`);
    console.log(`  Has XFA: ${info.hasXfa}`);

    console.log('\nMetadata:');
    if (info.metadata.title) console.log(`  Title: ${info.metadata.title}`);
    if (info.metadata.author) console.log(`  Author: ${info.metadata.author}`);
    if (info.metadata.subject) console.log(`  Subject: ${info.metadata.subject}`);
    if (info.metadata.creator) console.log(`  Creator: ${info.metadata.creator}`);
    if (info.metadata.producer) console.log(`  Producer: ${info.metadata.producer}`);

    console.log('\nPages:');
    info.pages.forEach((page) => {
      console.log(`  Page ${page.pageNumber}: ${page.width}x${page.height} (rotation: ${page.rotation}°)`);
    });
  });
}

/**
 * Example 12: Batch processing multiple PDFs
 */
async function example12_BatchProcessing() {
  const pdfFiles = ['doc1.pdf', 'doc2.pdf', 'doc3.pdf'];

  for (const file of pdfFiles) {
    console.log(`\nProcessing ${file}...`);

    // Extract text
    const text = await simple.extractText(file);
    await simple.saveTextToFile(file, `${file}.txt`);

    // Render first page
    await simple.renderToPNG(file, `${file}.png`, 0, 150);

    // Get info
    const pageCount = await simple.getPageCount(file);
    console.log(`  ${pageCount} pages processed`);
  }
}

/**
 * Example 13: Error handling
 */
async function example13_ErrorHandling() {
  try {
    // Try to open encrypted PDF without password
    const text = await simple.extractText('encrypted.pdf');
    console.log(text);
  } catch (error) {
    console.error('Error:', error.message);

    // Try again with password
    try {
      const text = await simple.extractText('encrypted.pdf', 'password123');
      console.log('Success with password!');
    } catch (error2) {
      console.error('Wrong password:', error2.message);
    }
  }
}

/**
 * Example 14: Using static helper functions
 */
async function example14_StaticHelpers() {
  // Static helper methods (same as simple API)
  const text = await EasyPDF.extractText('document.pdf');
  const buffer = await EasyPDF.renderPage('document.pdf', 0, { dpi: 300 });
  const info = await EasyPDF.getInfo('document.pdf');
  const results = await EasyPDF.search('document.pdf', 'important');

  console.log('Text length:', text.length);
  console.log('Buffer size:', buffer.length);
  console.log('Pages:', info.pageCount);
  console.log('Search results:', results.length);
}

/**
 * Example 15: Working with buffers (in-memory PDFs)
 */
async function example15_BufferOperations() {
  // Load PDF from buffer
  const fs = await import('fs/promises');
  const pdfBuffer = await fs.readFile('document.pdf');

  const pdf = EasyPDF.fromBuffer(pdfBuffer);
  try {
    const text = pdf.extractAllText();
    const pageCount = pdf.pageCount;

    console.log(`Loaded ${pageCount} pages from buffer`);
    console.log(`Extracted ${text.length} characters`);
  } finally {
    pdf.close();
  }
}

// ============================================================================
// CORE API - Full Control (for comparison)
// ============================================================================

/**
 * Example 16: Core API (traditional way, for comparison)
 */
function example16_CoreAPI() {
  // This is the traditional way - more verbose but maximum control
  const doc = Document.open('document.pdf');

  try {
    const pageCount = doc.pageCount;

    for (let i = 0; i < pageCount; i++) {
      const page = doc.loadPage(i);

      try {
        const text = page.extractText();
        console.log(`Page ${i}:`, text.substring(0, 100));

        const bounds = page.bounds();
        console.log(`Size: ${bounds.width}x${bounds.height}`);
      } finally {
        page.drop();
      }
    }
  } finally {
    doc.close();
  }
}

// ============================================================================
// Run all examples
// ============================================================================

async function runAllExamples() {
  console.log('='.repeat(80));
  console.log('NanoPDF Easy API Examples');
  console.log('='.repeat(80));

  // Note: Most examples are commented out to avoid requiring actual PDF files
  // Uncomment and modify paths as needed for your use case

  // await example1_SimpleTextExtraction();
  // await example2_SimpleRendering();
  // await example3_SimpleInfo();
  // await example4_SimpleSearch();
  // await example5_SimpleConversion();
  // await example6_FluentAPI();
  // await example7_ManualResourceManagement();
  // await example8_AdvancedRendering();
  // await example9_StructuredText();
  // await example10_DetailedSearch();
  // await example11_CompleteInfo();
  // await example12_BatchProcessing();
  // await example13_ErrorHandling();
  // await example14_StaticHelpers();
  // await example15_BufferOperations();
  // await example16_CoreAPI();

  console.log('\nAll examples completed!');
}

// Uncomment to run:
// runAllExamples().catch(console.error);

export {
  example1_SimpleTextExtraction,
  example2_SimpleRendering,
  example3_SimpleInfo,
  example4_SimpleSearch,
  example5_SimpleConversion,
  example6_FluentAPI,
  example7_ManualResourceManagement,
  example8_AdvancedRendering,
  example9_StructuredText,
  example10_DetailedSearch,
  example11_CompleteInfo,
  example12_BatchProcessing,
  example13_ErrorHandling,
  example14_StaticHelpers,
  example15_BufferOperations,
  example16_CoreAPI
};

